webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala

```diff
@@ -1215,97 +1242,56 @@ object ConstructResponseUtilV2 {
+                          responderManager: ActorRef,
+                          targetSchema: ApiV2Schema,
+                          settings: SettingsImpl,
+                          requestingUser: UserADM)(implicit stringFormatter: StringFormatter, timeout: Timeout, executionContext: ExecutionContext): Future[ReadResourcesSequenceV2] = {
+
+        val visibleResourceIris: Seq[IRI] = orderByResourceIri.filter(resourceIri => mainResourcesAndValueRdfData.resources.keySet.contains(resourceIri))
 
         // iterate over orderByResourceIris and construct the response in the correct order
-        val readResourceFutures: Vector[Future[ReadResourceV2]] = orderByResourceIri.map {
+        val readResourceFutures: Vector[Future[ReadResourceV2]] = visibleResourceIris.map {
             resourceIri: IRI =>
+                constructReadResourceV2(
+                    resourceIri = resourceIri,
+                    resourceWithValueRdfData = mainResourcesAndValueRdfData.resources(resourceIri),
+                    mappings = mappings,
+                    queryStandoff = queryStandoff,
+                    versionDate = versionDate,
+                    responderManager = responderManager,
+                    targetSchema = targetSchema,
+                    settings = settings,
+                    requestingUser = requestingUser
+                )
 
-                // the user may not have the permissions to see the resource
-                // i.e. it may not be contained in searchResults
-                searchResults.get(resourceIri) match {
-                    case Some(assertions: ResourceWithValueRdfData) =>
-                        // sufficient permissions
-                        // add the resource to the list of results
-                        constructReadResourceV2(
-                            resourceIri = resourceIri,
-                            resourceWithValueRdfData = assertions,
-                            mappings = mappings,
-                            queryStandoff = queryStandoff,
-                            versionDate = None,
-                            responderManager = responderManager,
-                            targetSchema = targetSchema,
-                            settings = settings,
-                            requestingUser = requestingUser
-                        )
-
-                    case None =>
-                        // include the forbidden resource instead of skipping (the amount of results should be constant -> limit)
-                        Future(forbiddenResource.getOrElse(throw AssertionException(s"Not all resources are visible, but forbiddenResource is None")))
-
-                }
         }.toVector
 
-        Future.sequence(readResourceFutures)
+        for {
+            resources <- Future.sequence(readResourceFutures)
+            mayHaveMoreResults = calculateMayHaveMoreResults && orderByResourceIri.size == settings.v2ResultsPerPage
```

Comment:
shouldn't this be `visibleResourceIris.size == settings.v2ResultsPerPage`?

Reply:
No, the logic is like this: if the page was full **before** filtering for permissions, then there might be another page to request.

Reply:
I've done some further refactoring to make this clearer in #1626.

Reply:
And in this PR I've just added some more comments to make this clear (in 6e96460).

Reply:
perfect, thanks.